### PR TITLE
Graphsync Fetcher Tests

### LINKED ISSUE #3232: Extended graphsync-based chain sync tests
### Description
Chain sync is a security-critical function and needs robust testing. The new graphsync-based syncer is great, but we need more confidence in it.

Extend unit and/or integration tests to cover:

- [x] partial responses
    - [x] missing Messages/MessageReceipts collection
    - [x] missing one block in multi-block tip (in recursive fetch)
    - [x] stopping after every height in [1, 5] in a recursive fetch (to catch batch edges)
- [x] a Block decodes but is not syntactically valid
- [x] the structure pointed to by Block.Messages/MessageReceipts does not decode
- [x] the structure pointed to by Block.Messages/MessageReceipts is not syntactically valid (there is no validator for messages implemented yet)
- [x] more to come...

In order to make these tests concise, I expect it'll be worth some effort extending the existing mock structures for more power and clarity.

Note: when the message/receipt structures themselves become trees, we'll also need to test for partial response of those.

### Acceptance criteria
- the graphsync-based fetcher is thoroughly covered by tests

### Where to begin
Existing tests are a good start

@hannahhoward

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

Thoroughly test the Graphsync Fetcher 
fix #3232 

# Implementation

- Significantly enhancing mocking for Graphsync -- based on previous PR https://github.com/filecoin-project/go-filecoin/pull/3264 + improvements based on feedback
- Add initial sketch of a MessageSyntaxValidator (but no implementation)
- Add code to Graphsync fetcher to handle non-decodable Message/Receipt blocks and also to validate them
- Move methods for generating fake unique receipts to general testing_messages utility
- Add DecodeMessages / DecodeReceipts methods for quickly decoding messages and receipts form raw bytes (for now)
- Add tests for behavior in case of:
   - missing Messages/MessageReceipts collection
   - missing one block in multi-block tip (in recursive fetch)
   - stopping after every height in [1, 5] in a recursive fetch (to catch batch edges)
   - block decodes but is not syntactically valid
   - the structure pointed to by Block.Messages/MessageReceipts does not decode
   - the structure pointed to by Block.Messages/MessageReceipts is not syntactically valid (though only through mocking a validator that errors)

# Discussion

Should this also include a solution for #3312 ?

### CODE REVIEW

**Comment by anorth** [comment]
I haven't reviewed this yet, but it's giant. The notes indicate a number of refactors ahead of the actual meat of this PR: could you please pull them into a "prefactor" PR to land ahead of this one?

**Comment by hannahhoward** [comment]
Set base branch to feat/fetcher-prefetcher till that is merged.

File: net/graphsync_fetcher_test.go [hunk 1]
```diff
@@ -231,19 +310,224 @@ func TestGraphsyncFetcher(t *testing.T) {
...
+	})
+
+	t.Run("stopping at edge heights in recursive fetch", func(t *testing.T) {
+		gen := builder.NewGenesis()
+		recursive16stop := builder.BuildManyOn(1, gen, withMessageBuilder)
+		recursive16middle := builder.BuildManyOn(15, recursive16stop, withMessageBuilder)
+		recursive4stop := builder.BuildManyOn(1, recursive16middle, withMessageBuilder)
+		recursive4middle := builder.BuildManyOn(3, recursive4stop, withMessageBuilder)
+		recursive1stop := builder.BuildManyOn(1, recursive4middle, withMessageBuilder)
+		final := builder.BuildOn(recursive1stop, 1, withMessageEachBuilder)
```
Comment (anorth) [inline net/graphsync_fetcher_test.go]: The coupling of this test to the specific expansion strategy used by the fetcher worries me. It's very easy to imagine that code changing and this test becoming useless. Could we change it to loop and test every stopping point from 1 to 22?

File: node/node.go [hunk 2]
```diff
@@ -33,8 +33,8 @@ import (
 	"github.com/libp2p/go-libp2p-core/host"
 	p2pmetrics "github.com/libp2p/go-libp2p-core/metrics"
 	"github.com/libp2p/go-libp2p-core/routing"
-	"github.com/libp2p/go-libp2p-kad-dht"
-	"github.com/libp2p/go-libp2p-kad-dht/opts"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	dhtopts "github.com/libp2p/go-libp2p-kad-dht/opts"
```
Comment (anorth) [inline node/node.go]: Revert these unnecessary aliases

File: consensus/message_validation.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+type MessageSyntaxValidator interface {
+	ValidateMessagesSyntax(ctx context.Context, messages []*types.SignedMessage) error
+	ValidateReceiptsSyntax(ctx context.Context, receipts []*types.MessageReceipt) error
+}
+
+// TODO: Define a message semantics validator if necessary
+// See: https://github.com/filecoin-project/go-filecoin/issues/3312
+
+// DefaultMessageSyntaxValidator implements the MessageSyntaxValidator interface.
+type DefaultMessageSyntaxValidator struct {
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply consensus/message_validation.go]: @ZenGround0 curious for your feedback before I go do this.

[the remark this answers is not in the document]
Comment (ZenGround0) [reply consensus/message_validation.go]: No strong opinions.  Single concrete type SGTM

**Comment by anorth** [comment]
Still LGTM

**Pull request merged**